### ISSUE TITLE
log-style API proposal

### ISSUE THREAD

**Issue by kazimuth** [metrics-rs/metrics#18]
I'm excited people are working on a metrics library for rust! I've been wanting one for a while. Your reddit post said you were okay with big API design ideas, so here are some i've been noodling on :)

The design I was contemplating is something like [`log`](https://github.com/rust-lang-nursery/log), where there's a tiny facade crate with simple macros that library authors can use freely, that calls into a global trait object for functionality; binaries can then install an implementation that does what they want.

E.g., library code can do something like:

```rust
use metrics::gauge;

gauge!(request_time, 123);
```

(Or, if we want to get fancy, you could also more metadata, like prometheus-style tags and opencensus-style units:)
```rust
gauge!(request_time, 123ns, endpoint="/app/thing", method=http::POST);
```

You could also have other macros to record counters, histograms, time blocks of code, etc. But it's all just simple macros.

Then, users can install a metrics implementation:

```rust
fn main() {
   // note: NOT the core metrics crate!
   magic_metrics_library::setup()
     .with_recorder(PrometheusRecorder::new())
     .with_exporter(HttpExporter::new())
     .filter("hyper") // don't include hyper metrics...
     .install();

   start_http_server();
}
```

And boom, they're done! That's the whole user facing API. Users don't need to know anything more about the internal details of the library. This is especially nice because it means library authors can just add metrics, fire-and-forget style; the core crate is so small that it adds very little compile-time cost, and also practically no runtime cost. (`log` without a logger installed has a 1ns overhead per-call, by my measurements.) As a result, when a user installs a metric exporter, they'll hopefully have a load of metrics available for free; assuming the facade crate gains adoption.

Now, the implementation. Basically, the core metrics crate contains a core `Metrics` trait that allows metrics to be recorded -- the API surface looks similar to the `Recorder` trait, but its responsibilities are broader; it's responsible for the whole metric recording process. It also contains a global, atomically initialized `static METRICS: Box<dyn Metrics>` trait object. Then, each `gauge!`, `counter!`, or `histogram!` call simply calls a method on this trait object. (You can look at `log` to see how this works.)

Note that we don't even have SnapshotProvider / Exporter / etc here! Those are super useful, but they can be moved to a crate for backend implementors -- we don't want their compile time cost in the facade crate. (Note: I've been calling the facade crate `metrics` in the style of `log`, but it could be `metrics_core` instead.)

So, that's the design sketch. Thoughts? I know this is pretty significantly different from the current design of the crate, but no code would actually need to be deleted -- the current code would only need a thin shim to comply with this API. This would also allow users more flexibility -- e.g. library authors don't need to be worried about dragging crossbeam in as a dependency, since the facade crate wouldn't have any dependencies; users pay for what they need.

**Comment by kazimuth** [comment]
Notes on optimization: it might seem like this design is incompatible with the high-performance crossbeam-based design this crate currently uses. Doesn't that require local sinks for each thread? Sure, we can do that, with... thread-locals:

```rust
thread_local! {
   static SINK: Sink = GLOBAL_SINK.clone();
}
impl Metrics for SinkCreator {
   fn record_gauge(data: GaugeData) {
      SINK.with(|sink| sink.send(data))
   }
}
```

Another fun optimization is allowing the global Metrics implementation to pre-process callsites. Since most of the information about a single `gauge!` invocation won't change -- the metric type, unchanging tags, etc. -- you can allow memoizations like so:

```rust
pub struct CallsiteToken(usize);

// data about a callsite that doesn't change.
struct CallsiteMetadata {
    metric_type: MetricType,
    source_module: &str,
    tags: &[Tag],
    // ... other stuff ...
}

trait Metrics {
   fn init_callsite(metadata: &'static CallsiteMetadata) -> CallsiteToken;
   fn record(token: CallsiteToken, value: f64);
}
```

Now, associate each callsite with a static, atomically initialized CallsiteToken; that token will be initialized with the callsite's metadata, and then passed into `record` whenever that callsite is hit in the future. Since the token's just an int, and the implementation controls its value, the implementation can use it to directly index into tables of sample buffers, etc. This means you don't need to hash recordings and things at runtime :)

**Comment by tobz** [comment]
So, I'll have to take a deeper look at all of this to make sure I'm grokking your proposal properly, but first thoughts:

I love the idea of providing a path for "simplest possible metrics configuration" based on macros and a global receiver.  At some point, you basically want/need a global `Receiver` anyways to be able to grab a new sink from a given thread, and it helps make the ownership of the `Receiver` (since it needs to stick around for the export stuff) simpler.  Just make it a lazy_static, done.

With that said, a lot of the spirit of your proposal, IMO, aligns heavily with what `tokio-trace` is looking to do, which is nearly identical from the perspective of a facade crate that, as more people use it in their crates, end users get the instrumentation benefit for free*

And with _that_ said, the crux of why we created `metrics-core` with those foundational traits was the idea of being able to create a `tokio-trace` subscriber -- say, `tokio-trace-metrics`, that would take something that fit the "record some metrics" traits and be able to turn spans into metrics for us, for free*.  We'd be slightly abusing the distributed tracing-orientedness to derive these structured metrics for end users with little to no input, but it would get us a lot of timing-based things for free.  Since `tokio-trace` also supports free-form `event!` calls -- basically identical to `log!` and spitting out a single entry -- we figured we could also jam counters/gauges in there as well with some field name matching.

On top of that, the intention to support filtering of given metrics/spans, etc, is already a working idea for `tokio-trace`.  And, since it's likely to come up as a question for a long time: `tokio-trace` does not need `tokio` to run at all... it's only related in name, and can be used in any application, so it still aligns with `metrics` on that front.

With _alllll_ of that said, I'm still intrigued by `metrics` itself being able to provide this stuff for people who may not want `tokio-trace` or aren't yet tied into that ecosystem.  It's totally possible we may decide we're not OK with having `tokio-trace`-specific macros and instead want metrics-specific macros, etc... so I'm gonna leave this open to ruminate on. :)

**Comment by kazimuth** [comment]
Oh, I hadn't looked at tokio-trace in a long time, that makes sense! (they should really change their name to just, like, "trace", lol.)

I do think it makes sense to provide a metrics facade crate, even if that crate is implemented using tokio-trace. There are plenty of metrics that aren't solely time- or count-based, ideally there'd be a clean api for using those.

**Comment by LucioFranco** [comment]
I think the facade is a great idea as a standalone, for tokio-trace I already have a working subscriber for internal use at work. My hope is to open source this soonish.

**Comment by tobz** [comment]
Gonna close this ticket as I just merged #20 🎉 

There should probably be a follow-up issue to discuss static callsite metadata/tokenization, as that involves some trickier synchronization and needs some thought.
